### PR TITLE
Merged improvement by @roybass for parallel builds into DisplayProgressMojo

### PR DESCRIPTION
**Pull request by wborn**
Merged the fix by @roybass into the DisplayProgressMojo and it seems to work well with single and multi threaded builds. Tested it on a Maven build with 300+ modules. It would be nice if this ends up in Maven Central :-)

### CODE REVIEW

**Comment by roybass** [comment]
Great! Thanks for merging!

**Comment by wokier** [comment]
Partially integrated in v0.6
I kept the minimumNotifyMillis option
Released in Maven Central

File: src/main/java/com/github/wokier/maven/progress/NotifyProgressMojo.java [hunk 1]
```diff
@@ -9,63 +8,54 @@
...
-      final int reactorProjectsCount = reactorProjects.size();
-      final File currentProjectBasedir = currentProject.getBasedir();
-      final String message = ProgressUtils.progress(projectIndex, reactorProjectsCount, currentProjectBasedir);
-
-      ExecutorService singleThreadExecutor = Executors.newSingleThreadExecutor();
-      singleThreadExecutor.execute(new Runnable()
-      {
-         public void run()
-         {
-            Notify.getInstance().notify(MessageType.INFO, "Reactor Progress", message);
```
Comment (wokier) [inline src/main/java/com/github/wokier/maven/progress/NotifyProgressMojo.java:54]: Why is this executed in a new thread ?

File: src/main/java/com/github/wokier/maven/progress/ProgressUpdater.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.channels.OverlappingFileLockException;
+import java.util.Date;
+
+import org.apache.maven.plugin.MojoExecutionException;
+
+/**
+ * Updates the progress of a reactor by storing and retrieving the progress from a properties file.
+ * Because the plugin can run multithreaded in several classloaders a file is used to record state.
```
Comment (wokier) [inline src/main/java/com/github/wokier/maven/progress/ProgressUpdater.java:16]: What are the maven options to use different class loaders in parallel builds ?

**Comment by wokier** [comment]
Thanks for your contribution. I just added some questions on parts i did not well understood.

File: src/main/java/com/github/wokier/maven/progress/NotifyProgressMojo.java (hunk 1 above)
Comment (wborn) [inline src/main/java/com/github/wokier/maven/progress/NotifyProgressMojo.java:54]: The creation of the new thread was already there when I forked the code. Here is a link to the commit in which you added it: https://github.com/wokier/progress-maven-plugin/commit/c0dc5e4cfede81fb2e447496cc58bf3bebbce1e4

I was wondering myself what the reason would be. What I can imagine is that you do not want the Maven build to wait on the notification subsystem. However I did not look into the implementation of Notify and see if a new thread is created there anyway.

File: src/main/java/com/github/wokier/maven/progress/ProgressUpdater.java (hunk 2 above)
Comment (wborn) [inline src/main/java/com/github/wokier/maven/progress/ProgressUpdater.java:16]: I am building with Maven 3.x with the -T 1C option. 

Here is a link that describes when Maven uses several ClassLoaders: http://takari.io/book/91-maven-classloading.html

So looks like it depends on the plugins you use if you are using them with different configurations around your projects.

I encountered it while building some projects with:
- flexmojos-maven-plugin and packaging=swf in the POM
- android-maven-plugin and packaging=apk in the POM

File: src/main/java/com/github/wokier/maven/progress/NotifyProgressMojo.java (hunk 1 above)
Comment (wokier) [inline src/main/java/com/github/wokier/maven/progress/NotifyProgressMojo.java:54]: So it looks like i have done it and forget ...
If there is performance or reliability issues, that should be handled in the lower layer java-to-OS-notify. That's why i have done distinct projects.

**Pull request closed without merging**